### ISSUE TITLE
Jax hangs when using PositionalSharding and fft2.

### ISSUE THREAD

**Issue by nithinvc** [google/jax#17404]
### Description

I would like to use positional sharding to batch some computation (involving fft2) across gpus. The program hangs at jax.numpy.fft.fft2. As a minimal example:

```python
import jax
import jax.numpy as jnp
from jax.experimental import mesh_utils
from jax.sharding import PositionalSharding
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (8, 64, 64, 64, 2))
@jax.jit
def f(x):
    y = jnp.fft.fft2(x, axes=(0, 1))
    return y    

n_devices = jax.local_device_count()
sharding = PositionalSharding(mesh_utils.create_device_mesh(n_devices,)).reshape((n_devices, 1, 1, 1, 1))
_x = jax.device_put(x, sharding)

# This works fine
print('Non-sharded', jax.vmap(f)(x).shape)

# This does not
jax.vmap(f)(_x)
```

Weirdly, the GPU utilization stays at 100% with no deviation. 

### What jax/jaxlib version are you using?

jax v0.4.14, jaxlib v0.4.14+cuda12.cudnn89

### Which accelerator(s) are you using?

GPU

### Additional system info

Python 3.10, Ubuntu 22.04.2 LTS, cuDNN v8.9.4

### NVIDIA GPU info

```
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 530.30.02              Driver Version: 530.30.02    CUDA Version: 12.1     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                  Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf            Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA RTX A6000                On | 00000000:4F:00.0 Off |                  Off |
| 30%   37C    P2               91W / 300W|   2855MiB / 49140MiB |    100%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA RTX A6000                On | 00000000:52:00.0 Off |                  Off |
| 30%   42C    P2               98W / 300W|    915MiB / 49140MiB |    100%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   2  NVIDIA RTX A6000                On | 00000000:53:00.0 Off |                  Off |
| 30%   40C    P2               76W / 300W|   4955MiB / 49140MiB |     39%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   3  NVIDIA RTX A6000                On | 00000000:56:00.0 Off |                  Off |
| 33%   61C    P2              208W / 300W|  38695MiB / 49140MiB |     99%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   4  NVIDIA RTX A6000                On | 00000000:57:00.0 Off |                  Off |
| 35%   65C    P2              211W / 300W|  38695MiB / 49140MiB |      7%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   5  NVIDIA RTX A6000                On | 00000000:CE:00.0 Off |                  Off |
| 36%   64C    P2              200W / 300W|  38695MiB / 49140MiB |     99%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   6  NVIDIA RTX A6000                On | 00000000:D1:00.0 Off |                  Off |
| 49%   73C    P2              287W / 300W|  37403MiB / 49140MiB |     94%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   7  NVIDIA RTX A6000                On | 00000000:D2:00.0 Off |                  Off |
| 33%   63C    P2              188W / 300W|  48431MiB / 49140MiB |     99%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   8  NVIDIA RTX A6000                On | 00000000:D5:00.0 Off |                  Off |
| 30%   26C    P8               13W / 300W|  10097MiB / 49140MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
|   9  NVIDIA RTX A6000                On | 00000000:D6:00.0 Off |                  Off |
| 30%   29C    P8               20W / 300W|    663MiB / 49140MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
```

**Comment by erfanzar** [comment]
If you trying to Shard your array across GPU/CPUor TPU or any other accelerator that you have the best and most efficient way to do is using pjit and I recommend you to use 
```python

import os
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = 'false'
os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = 'gpu'

```

and if you want to use pjit the easiest and fastest method for your code is 
```python

import jax
import jax.numpy as jnp
from jax.experimental import mesh_utils, pjit

from jax.sharding import Mesh, PartitionSpec as PS
from fjutils import with_sharding_constraint  # smarter and more stable version 
# or
from jax.experimental.pjit import with_sharding_constraint

key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (8, 64, 64, 64, 2))


@jax.jit
def f(x):
    y = jnp.fft.fft2(x, axes=(0, 1))
    return y


n_devices = jax.local_device_count()
mesh = mesh_utils.create_device_mesh(
    jnp.ones((1, n_devices)).shape, ('dp', 'fsdp', 'mp')
)

with mesh:
    x = with_sharding_constraint(
        x, PS('fsdp')  # the array shape is (8, 64, 64, 64, 2) and you have 10 devices this won't work
        # I recommend to set visible devices to 8 or reshape the array like (n_devices,-1,2)
    )
    res = pjit.pjit(
        f,
        in_shardings=PS(),
        out_shardings=PS()
    )(x)
    print(x)
```

and the reason that this code of your
```python
jax.vmap(f)(_x)
```
is not working because you have to pass the full array variable name to the func in case you are using the single machine I guess

**Comment by nithinvc** [comment]
Hi! Thanks for the response!

Sorry, I should clarify. We've ensured that when batching, the size of the large batch always evenly divides the number of devices (doing something like .reshape(n_devices, -1,...). Even when restricted to the case of 2 GPUs, the code I've listed in my original post fails. 

I tried running the code that you've provided and have a few questions: 

```python
mesh = mesh_utils.create_device_mesh(
    jnp.ones((1, n_devices)).shape, ('dp', 'fsdp', 'mp')
)
```
This line produces the following error: 
```python
AttributeError: 'str' object has no attribute 'device_kind'
```
Q1: The documentation for create_device_mesh states lists the second argument as "devices: optionally, the devices to construct a mesh for. Defaults to jax.devices()." It seems like this argument should be the GPU or TPU object representing the accelerator? Can you elaborate on what the string args are doing? 

Q2: Is there a difference between pjit and jit? The jax documentation has the line "NOTE: This function is now equivalent to jax.jit please use that instead" so I'm a little confused on why this would have different behavior vs. jit. 

Q3: What does mesh as a context manager do in this case? Does it automatically apply jax.device_put to arrays created within the context block? When I try to use the resulting mesh as a context manager, an attribute error is thrown:
```python
AttributeError: __enter__
```
Removing the context manager does throw the following error:
```python
RuntimeError: pjit requires a non-empty mesh if you are passing `PartitionSpec`s or `None` to in_shardings!
```

Q4: You mentioned vmap(f)(_x) is not working because
> you have to pass the full array variable name to the func in case you are using the single machine

I'm not sure I understand. _x should be sharded after using device_put (to my understanding from the [tutorial notebook](https://jax.readthedocs.io/en/latest/notebooks/Distributed_arrays_and_automatic_parallelization.html)). I thought the vmap takes into account the input argument sharding?

Regardless, in order to make the code semi-run, I had to make a few changes:

```python
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '6,7' # Restricted to 2 GPUs for simplicity
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = 'false'

# Commented out - Including this line prevents jax from seeing the GPUs? Throws an error "falling back to CPU"
# os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = 'gpu'

import jax
import jax.numpy as jnp
from jax.sharding import NamedSharding, PositionalSharding, Mesh
from jax.experimental import mesh_utils, pjit
from jax.sharding import Mesh, PartitionSpec as PS
from fjutils import with_sharding_constraint

n_devices = jax.local_device_count()
sharding = PositionalSharding(mesh_utils.create_device_mesh(n_devices,)).reshape((n_devices, 1, 1, 1, 1)) 
key = jax.random.PRNGKey(0)
x = jax.random.normal(key, (2, 64, 64, 64, 2))

@jax.jit
def f(x):
    y = jnp.fft.fft2(x, axes=(0, 1))
    return y

devices = mesh_utils.create_device_mesh(
    jnp.ones((1, n_devices, 1)).shape,
)
mesh = Mesh(devices, ('dp', 'fsdp', 'mp'))

with mesh:
    x = with_sharding_constraint(
        x, PS('fsdp')
    )
    res = pjit.pjit(
        f,
        in_shardings=PS(),
        out_shardings=PS()
    )(x)
    print(x)
```
Which results in the following error:
```python 
ValueError: Sharding passed to pjit does not match the sharding on the respective arg. Got pjit sharding: NamedSharding(mesh={'dp': 1, 'fsdp': 2, 'mp': 1}, spec=PartitionSpec()),
arg sharding: NamedSharding(mesh={'dp': 1, 'fsdp': 2, 'mp': 1}, spec=PartitionSpec('fsdp',)) for arg shape: (2, 64, 64, 64, 2)
```
Any advice is appreciated. Thanks!

**Comment by erfanzar** [comment]
sorry for wrong code this one will work 

```python
import os

os.environ['CUDA_VISIBLE_DEVICES'] = '6,7'  # Restricted to 2 GPUs for simplicity
os.environ['XLA_PYTHON_CLIENT_PREALLOCATE'] = 'false'

# Commented out - Including this line prevents jax from seeing the GPUs? Throws an error "falling back to CPU"
# os.environ['XLA_PYTHON_CLIENT_ALLOCATOR'] = 'gpu'

import jax
import jax.numpy as jnp
from jax.sharding import NamedSharding, PositionalSharding, Mesh
from jax.experimental import mesh_utils, pjit
from jax.sharding import Mesh, PartitionSpec as PS
from fjutils import with_sharding_constraint


def main():
    n_devices = jax.local_device_count()
    sharding = PositionalSharding(mesh_utils.create_device_mesh(n_devices, )).reshape((n_devices, 1, 1, 1, 1))
    key = jax.random.PRNGKey(0)
    x = jax.random.normal(key, (2, 64, 64, 64, 2))

    devices = mesh_utils.create_device_mesh(
        jnp.ones((1, n_devices, 1)).shape,
    )
    mesh = Mesh(devices, ('dp', 'fsdp', 'mp'))

    with mesh:
        def f(x):
            x = with_sharding_constraint(
                x, PS('fsdp')
            )
            y = jnp.fft.fft2(x, axes=(0, 1))
            return y

        res = pjit.pjit(
            f,
            in_shardings=PS(),
            out_shardings=PS()
        )(x)
        print(res)


if __name__ == "__main__":
    main()

```

## Q1 Answer
for using `jax.devices()` you have to pass the backend name that  you want to get the list of device of that for example in your case you have 10 RTX A6000 if you use `jax.devices('gpu')` you will get list of your gpus their position and ... but if you try to get tpu it will raise an error cause you don't have any tpu on your machine `jax.devices('tpu')` and if you want to have a use case example imaging you want to use offload method in flax to load params on CPU and then shard them across all of the devices you can use

```python

with jax.default_device(jax.devices('cpu')[0]):
    .... # load your params here
```

## Q2 Answer

Yes, there is a difference between pjit and jit. pjit is a deprecated function that was used to compile functions for parallel execution on multiple devices. jit is the newer function that should be used instead.

The main difference between pjit and jit is that pjit requires the user to specify the sharding of the input and output arguments. Sharding is a way of dividing up a computation across multiple devices. The sharding of an argument specifies how many times it should be replicated across the devices.

For example, if you have a function that takes an input array of size 100 and you want to run it on 2 devices, you would need to specify the sharding of the input array as (2, 50). This means that the input array would be replicated 2 times across the devices, with each device getting a slice of size 50.

The sharding of the output arguments is also important because it determines how the outputs of the function are distributed across the devices. If you do not specify the sharding of the output arguments, then jit will try to infer the sharding automatically. However, this may not always be possible, and it can lead to suboptimal performance.

For these reasons, it is generally recommended to use jit instead of pjit. jit is more flexible and easier to use, and it does not require the user to specify the sharding of the input and output arguments.

also keep in mind that pjit is still in the experimental state


## Q3 Answer
pjit required to be used in mesh

**Comment by nithinvc** [comment]
Hi, thanks for your response! The code that you provided still does not run. Crucially, it hangs on lines:

```python
        res = pjit.pjit(
            f,
            in_shardings=PS(),
            out_shardings=PS()
        )(x)
```
And never prints ```res```. Similar to the script that I posted, the GPU utilization peaks at 100% and hangs. Even giving it sufficient time to complete (~1hr) yields no difference. Do you have a suggestion for how I would go about finding the source of the hang?

**Comment by erfanzar** [comment]
> Hi, thanks for your response! The code that you provided still does not run. Crucially, it hangs on lines:
> 
> ```python
>         res = pjit.pjit(
>             f,
>             in_shardings=PS(),
>             out_shardings=PS()
>         )(x)
> ```
> 
> And never prints `res`. Similar to the script that I posted, the GPU utilization peaks at 100% and hangs. Even giving it sufficient time to complete (~1hr) yields no difference. Do you have a suggestion for how I would go about finding the source of the hang?

actually i ran this code twice on Kaggle TPUs and GPUs and the code worked on both of them
and this was output

```text
[[[[[ 3.24833870e+00+0.00000000e+00j -1.67170668e+00+0.00000000e+00j]
    [ 1.13448925e+01+0.00000000e+00j -1.56441584e+01+0.00000000e+00j]
    [-2.06614799e+01+0.00000000e+00j -4.72019482e+00+0.00000000e+00j]
    ...
    [-8.73501062e-01+0.00000000e+00j  1.61715186e+00+0.00000000e+00j]
    [-6.35988188e+00+0.00000000e+00j  4.83073378e+00+0.00000000e+00j]
    [ 6.58679754e-02+0.00000000e+00j  1.09692688e+01+0.00000000e+00j]]

   [[ 1.50113926e+01+0.00000000e+00j  1.85722128e-01+0.00000000e+00j]
    [ 1.44338236e+01+0.00000000e+00j  1.20921860e+01+0.00000000e+00j]
    [-2.31441736e+00+0.00000000e+00j -2.15576148e+00+0.00000000e+00j]
    ...
    [-5.78798234e-01+0.00000000e+00j -1.50790095e+00+0.00000000e+00j]
    [-8.80465508e+00+0.00000000e+00j  2.43676052e+01+0.00000000e+00j]
    [ 1.34009323e+01+0.00000000e+00j  2.30402994e+00+0.00000000e+00j]]
```

**Comment by hawkinsp** [comment]
This seems fixed to me with jax/jaxlib 0.4.20. Closing, on the belief that it's fixed. If it's still a problem with an up to date JAX, leave a comment and I'll reopen the issue. Hope that helps!
